lib/src/migrators/module.dart

```diff
@@ -318,64 +319,70 @@ class _ModuleMigrationVisitor extends MigrationVisitor {
   @override
   void visitStylesheet(Stylesheet node) {
     var oldNamespaces = _namespaces;
-    var oldHasUseRule = _hasUseRule;
+    var oldHasUseRule = _usedUrls;
     var oldAdditionalUseRules = _additionalUseRules;
     var oldUseAllowed = _useAllowed;
-    _namespaces = {};
-    _hasUseRule = {};
+    _usedUrls = {};
     _additionalUseRules = {};
     _useAllowed = true;
-    _determineNamespaces(node.span.sourceUrl);
+    _namespaces = _determineNamespaces(node.span.sourceUrl);
     super.visitStylesheet(node);
     _namespaces = oldNamespaces;
-    _hasUseRule = oldHasUseRule;
+    _usedUrls = oldHasUseRule;
     _additionalUseRules = oldAdditionalUseRules;
     _lastUrl = node.span.sourceUrl;
     _useAllowed = oldUseAllowed;
   }
 
   /// Determines namespaces for all `@use` rules that the stylesheet will
   /// contain after migration.
-  void _determineNamespaces(Uri url) {
+  Map<Uri, String> _determineNamespaces(Uri url) {
+    var namespaces = <Uri, String>{};
     var sourcesByNamespace = <String, Set<ReferenceSource>>{};
-    for (var reference in references.referenceSources.keys) {
-      if (reference.span.sourceUrl == url) {
-        var source = references.referenceSources[reference];
-        var namespace = source.defaultNamespace;
-        if (namespace != null) {
-          sourcesByNamespace.putIfAbsent(namespace, () => {}).add(source);
-        }
+    for (var reference in references.sources.keys) {
+      if (reference.span.sourceUrl != url) continue;
+      var source = references.sources[reference];
+      var namespace = source.defaultNamespace;
+      if (namespace == null) continue;
+      // Existing `@use` rules should always keep their namespaces.
+      if (source is UseSource) {
+        namespaces[source.url] = namespace;
+      } else {
+        sourcesByNamespace.putIfAbsent(namespace, () => {}).add(source);
       }
     }
 
-    for (var namespace in sourcesByNamespace.keys) {
-      var sources = sourcesByNamespace[namespace];
-      if (sources.length == 1) {
-        _namespaces[sources.first.url] = namespace;
+    // First assign namespaces to those without conflicts.
+    var conflictingNamespaces = <String, Set<ReferenceSource>>{};
+    sourcesByNamespace.forEach((namespace, sources) {
+      if (sources.length == 1 && !namespaces.containsValue(namespace)) {
+        namespaces[sources.first.url] = namespace;
       } else {
-        _resolveNamespaceConflict(namespace, sources);
+        conflictingNamespaces[namespace] = sources;
       }
-    }
+    });
+
+    // Then resolve conflicts where they exist.
+    conflictingNamespaces.forEach((namespace, sources) {
+      if (sources.length > 1) {
```

Comment:
What happens if this conditional doesn't match? It looks like a namespace will just not be assigned, which seems wrong.

Reply:
Yeah, that conditional is not supposed to be present. I fixed it in one of the downstream PRs but forgot to fix it here.